### PR TITLE
Fine grained dynamic shape controls

### PR DESCRIPTION
**Pull request by voznesenskym**
https://docs.google.com/document/d/1aoIyYE8_6cYpWqS25thzVoIiKsT5aaUEOiiPwbIXt8k/edit

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -1038,38 +1047,14 @@ def create_symbol(self, val: int, source: Source) -> "sympy.Expr":
...
+        else:
             sympy_expr = Symbol(f"s{len(self.var_to_val)}", positive=True, integer=True)
-            self.var_to_val[sympy_expr] = sympy.Integer(val)
-            self.val_to_var[val] = sympy_expr
-
-        # This implements duck-shaping: input sizes that match are assigned
-        # the same symint
-        r = self.duck_int(val)
-        if isinstance(r, Symbol):
-            r.sources.append(source)
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1055]: Put back - used for debugging, and add a test

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -1038,38 +1047,14 @@ def create_symbol(self, val: int, source: Source) -> "sympy.Expr":
...
-        # Now attempt to duck size this value
-        # TODO: Use site has to duck size
-        # TODO: Do this duck sizing lazily later
-
-        # Create a duck sized int if necessary
-        if val not in self.val_to_var:
+        if val in self.val_to_var: 
+            # 0 or 1
+            sympy_expr = self.val_to_var[val]
+        else:
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1053]: Move into its own PR

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1178,6 +1164,13 @@ def track_symint(source, val):
             else:
                 input_guards.append((source, sympy.Integer(val)))
 
+        def _verify(expr, potential_expr):
+            for e in expr.free_symbols:
+                srcs = symbol_to_source[e]
+                for src in srcs:
+                    if src in dynamic_sources:
+                        raise RuntimeError(f"Attempting to introduce a guard {potential_expr} that violates user's mark_dynamic")
+
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1173]: Relax to match spec - expr.free_symbols == 1 --> assert. Otherwise, relational.

File: test/dynamo/test_dynamic_shapes.py [hunk 4]
```diff
@@ -25,19 +25,78 @@
...
-DynamicShapesReproTests = make_dynamic_cls(test_repros.ReproTests)
-DynamicShapesNNModuleTests = make_dynamic_cls(test_modules.NNModuleTests)
-DynamicShapesUnspecTests = make_dynamic_cls(test_unspec.UnspecTests)
-DynamicShapesExportTests = make_dynamic_cls(test_export.ExportTests)
-DynamicShapesSubGraphTests = make_dynamic_cls(test_subgraphs.SubGraphTests)
+DynamicShapesFunctionTests = make_dynamic_cls(
+    test_functions.FunctionTests, assume_static_by_default=False
+)
+DynamicShapesFunctionTestsDefaultStatic = make_dynamic_cls(
+    test_functions.FunctionTests, assume_static_by_default=True
```
Comment (voznesenskym) [inline test/dynamo/test_dynamic_shapes.py:45]: for assume_static_by_default in [True, False]

**Review by ezyang — COMMENTED** [review]
Looks good.

Things to do next:

1. Remove duck sizing in its own PR and validate perf / trace time consequences
2. 0/1 specialization is a potential soundness problem, need to talk to people about it
3. Seriously consider adding fast fail in this PR directly (or if you don't, print the backtraces when failing late)

**Comment by voznesenskym** [comment]
> Looks good.
> 
> Things to do next:
> 
> 1. Remove duck sizing in its own PR and validate perf / trace time consequences

Done, but we probably don't need this with the changes I made here now.

> 2. 0/1 specialization is a potential soundness problem, need to talk to people about it

Discussed offline

> 3. Seriously consider adding fast fail in this PR directly (or if you don't, print the backtraces when failing late)

Fast fail I think is going to take extra plumbing - if I print backtraces, which backtraces would you like to see?

File: test/dynamo/test_dynamic_shapes.py [hunk 5]
```diff
@@ -25,21 +25,83 @@
...
-DynamicShapesMiscTests = make_dynamic_cls(test_misc.MiscTests)
-DynamicShapesReproTests = make_dynamic_cls(test_repros.ReproTests)
-DynamicShapesNNModuleTests = make_dynamic_cls(test_modules.NNModuleTests)
-DynamicShapesUnspecTests = make_dynamic_cls(test_unspec.UnspecTests)
-DynamicShapesExportTests = make_dynamic_cls(test_export.ExportTests)
-DynamicShapesSubGraphTests = make_dynamic_cls(test_subgraphs.SubGraphTests)
+# Note - this is a little redundant, but some of these tests have specific
+# exclusions.
+# Note #2 - Putting this in a for loop over [True, False] for
+# assume_static_by_default clobbers the name, which makes the tests not run.
```
Comment (ezyang) [inline test/dynamo/test_dynamic_shapes.py:44]: You could do setattr(sys.module[__name__], blah) to programatically set the name

File: test/dynamo/test_export.py [hunk 6]
```diff
@@ -1794,6 +1799,40 @@ def forward(self, x):
...
+                return x.sin()
+            return x.cos()
+
+        torch._dynamo.export(my_dyn_fn, y)
+        torch._dynamo.mark_dynamic(y, 0)
+
+        with self.assertRaises(
+            torch._dynamo.exc.InternalTorchDynamoError,
+        ):
+            torch._dynamo.export(my_dyn_fn, y)
```
Comment (ezyang) [inline test/dynamo/test_export.py:1834]: Some more stuff to test:

- Relational guards
- Multiple dynamic dims
- 0/1 handling

File: test/dynamo/test_misc.py [hunk 7]
```diff
@@ -4273,6 +4286,129 @@ def dummy_fn():
...
+        y = torch.randn([3, 3, 3])
+
+        def my_dyn_fn(x):
+            if x.shape[0] > 3:
+                return x.sin()
+            return x.cos()
+
+        torch._dynamo.optimize("eager")(my_dyn_fn)(y)
+        torch._dynamo.mark_dynamic(y, 0)
+        torch._dynamo.reset()
```
Comment (ezyang) [inline test/dynamo/test_misc.py:4318]: Hmph. The reset here is a bit irritating. Maybe we should guard on the dynamic markers on the tensors.

File: test/dynamo/test_subgraphs.py [hunk 8]
```diff
@@ -379,10 +382,27 @@ def fn(a, b):
...
+            # of breaks as without dynamic, since no tensors were marked dyn.
+            self.assertEqual(cnt_dynamic.frame_count, steps)
+
+            torch._dynamo.reset()
+            # Reset the counter
+            cnt_dynamic = torch._dynamo.testing.CompileCounter()
+            opt_fn = torch._dynamo.optimize(cnt_dynamic, dynamic=False)(fn)
+            for i in range(start, end):
+                opt_fn(torch.randn(i), torch.randn(i))
+            self.assertEqual(cnt_dynamic.frame_count, steps)
```
Comment (ezyang) [inline test/dynamo/test_subgraphs.py:402]: What's this testing?

File: torch/_dynamo/config.py [hunk 9]
```diff
@@ -60,6 +60,10 @@
 # don't specialize on shapes and strides and put shape ops in graph
 dynamic_shapes = os.environ.get("TORCHDYNAMO_DYNAMIC_SHAPES") == "1"
 
+# This flag temporarily changes the behavior of dynamic_shapes=True.
```
Comment (voznesenskym) [inline torch/_dynamo/config.py:63]: This is written weird - it doesn't "temporarily change", rather, it's a temporary flag...

File: torch/_dynamo/__init__.py [hunk 10]
```diff
@@ -112,3 +112,49 @@ def fn(a):
 def graph_break():
     """Force a graph break"""
     pass
+
+
+def forbid_in_graph(fn):
+    """
+    Customize which functions TorchDynamo will assert are not present while tracing.
```
Comment (ezyang) [inline torch/_dynamo/__init__.py:119]: I don't think this really makes it clear. Maybe this extra text would help. "In particular, if you call a `forbid_in_graph` function inside a `torch.compile` region, this will result in a hard error and indicate this function call should be moved outside of the region." Which now makes me wonder, if you are in a torch.compile region but you happen to be in a frame that is not being compiled, should we raise anyway? This seems like a good idea in case we improve dynamo and now your frame is being compiled.

Comment (voznesenskym) [reply torch/_dynamo/__init__.py:119]: We should only raise when we hit upon this value in tracing - there is lots of prod code that sets up the analogue of `if tracing / if not tracing` and uses it to safeguard regions from dynamo access or to change behavior during tracing. The prior art for this, graph breaks, only occur when you happen upon it. If a frame with a graph break in it is not being compiled, for whatever reason, we do not graph break (in export, we do not error). I think this should match.

File: test/dynamo/test_dynamic_shapes.py (hunk 5 above)
Comment (voznesenskym) [reply test/dynamo/test_dynamic_shapes.py:44]: Is it worth it, at that point? no strong opinion, this felt the most readable atm.

File: torch/_dynamo/__init__.py [hunk 11]
```diff
@@ -112,3 +112,49 @@ def fn(a):
...
+
+
+@forbid_in_graph
+def mark_dynamic(t, index):
+    """
+    Mark a tensor as having a dynamic dim.
+
+    The behavior of having a dynamic dimension on a tensor is governed by a few factors:
+    1) torch._dynamo.config dynamic_shapes True or False - we do not support False atm, but will in the future.
+    dynamic_shapes must be True for mark_dynamic to work.
```
Comment (ezyang) [inline torch/_dynamo/__init__.py:139]: The phrasing of this suggests we should raise an error if you mark dynamic and dynamic shapes is not True. But this is not the case; I can mark without the config and then pass it to dynamic=True region no-problemo. So seems like it needs some rewording.

There's another question, which is if a NON dynamic torch.compile region gets a tensor marked as having dynamic dim, what should we do? Erroring here seems good

File: torch/_dynamo/config.py [hunk 12]
```diff
@@ -60,6 +60,10 @@
...
+# When assume_static_by_default is True, we only allocate symbols for shapes marked dynamic via mark_dynamic.
```
Comment (ezyang) [inline torch/_dynamo/config.py:64]: I think this will die before we move on and forget, but a good thing to do is mention under what condition this knob can be removed

File: torch/_dynamo/source.py [hunk 13]
```diff
@@ -178,6 +178,18 @@ def name(self):
         else:
             raise AssertionError(f"unhandled {self.prop}")
 
+    def __eq__(self, other):
+        return (
+            self.base == other.base
+            and self.prop == other.prop
+            and self.idx == other.idx
+        )
```
Comment (ezyang) [inline torch/_dynamo/source.py:186]: I am confuse; dataclasses automatically get equality defined on them, why you need

File: torch/_dynamo/source.py [hunk 14]
```diff
@@ -178,6 +178,18 @@ def name(self):
...
+
+    # TODO(voz): This isnt really used in any way where we care about having a real hash yet.
+    # This is just to be able to chuck these things in a set.
+    def __hash__(self):
+        return id(self)
```
Comment (ezyang) [inline torch/_dynamo/source.py:191]: This seems bad. If we really want hash, I'd suggest we start marking these Sources as frozen=True so they get hashes automatically generated. Best to do separately.

Comment (voznesenskym) [reply torch/_dynamo/source.py:191]: We could also use a list, there is not a ton of these :)

File: torch/_dynamo/variables/builder.py [hunk 15]
```diff
@@ -231,6 +231,10 @@ def make_guards(self, *guards):
     def _wrap(self, value):
         from ..comptime import comptime
 
+        if callable(value):
+            if hasattr(value, "_dynamo_forbidden") and value._dynamo_forbidden:
+                raise AssertionError(f"Attempt to trace forbidden callable {value}")
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:236]: This seems awkward? It suggests that if I merely reference a forbidden function such that it gets a tracker, but I don't call it, it will still result in an error. That seems bad.

File: torch/_dynamo/__init__.py (hunk 11 above)
Comment (voznesenskym) [reply torch/_dynamo/__init__.py:139]: > The phrasing of this suggests we should raise an error if you mark dynamic and dynamic shapes is not True. But this is not the case; 
> There's another question, which is if a NON dynamic torch.compile region gets a tensor marked as having dynamic dim, what should we do? Erroring here seems good

Is that not the test `test_no_dynamic_shapes_mark_dynamic_illegal` ? Perhaps I am misunderstanding what you mean? Can you elaborate?

File: torch/_dynamo/variables/builder.py (hunk 15 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:236]: Moved

File: torch/_dynamo/__init__.py (hunk 10 above)
Comment (voznesenskym) [reply torch/_dynamo/__init__.py:119]: I added some tests clarifying this - lmk if they align with what you expect.

File: test/dynamo/test_misc.py (hunk 7 above)
Comment (voznesenskym) [reply test/dynamo/test_misc.py:4318]: Oh interesting suggestion - but I kind of like the idea of these being counter to guards - in that they alleviate them. There is a lot of value in keeping them separate, I think. I think the idea here would be that we do not really have the data anymore of what was and was not marked dynamic - for all cases. However, if we want, this would be a very easy guard to add :)

File: test/dynamo/test_export.py (hunk 6 above)
Comment (voznesenskym) [reply test/dynamo/test_export.py:1834]: Added most of the above - what do you want to see tested in 0/1?

File: test/dynamo/test_dynamic_shapes.py (hunk 5 above)
Comment (ezyang) [reply test/dynamo/test_dynamic_shapes.py:44]: My preference is still for looping, because whenever a new test file is added, in the current configuration you have to do it twice

File: test/dynamo/test_export.py (hunk 6 above)
Comment (ezyang) [reply test/dynamo/test_export.py:1834]: This is not so much a "test" as it is "documenting the behavior". If I mark a dimension as dynamic, do I assume it is non-zero/non-one? I think it is best if we do NOT make this negative inference. This is more code though.

Comment (ezyang) [reply test/dynamo/test_export.py:1834]: You can test it by just having a condition on the dynamic dim that it is != 1, and having this appropriately throw?

File: test/dynamo/test_misc.py (hunk 7 above)
Comment (ezyang) [reply test/dynamo/test_misc.py:4318]: My main concern is a benchmarking one: if you warm up the cache with a static compile, then mark a dimension as dynamic and run with the same size, you will hit the static (and not recompile with dynamic shapes).

File: torch/_dynamo/__init__.py (hunk 10 above)
Comment (ezyang) [reply torch/_dynamo/__init__.py:119]: No, I don't think the analogy to graph breaks is applicable. Graph breaks never cause code to raise an error; it's always semantics preserving to add them, no matter if they do something or not. But forbidden operators CAN cause code to raise an error. I am saying it is more uniform and easier to predict if they raise an error if they ALWAYS raise an error based on some simple condition (is torch.compile in my stack) as opposed to a complicated one (did I happen to be processing this frame or not.)

File: torch/_dynamo/__init__.py (hunk 11 above)
Comment (ezyang) [reply torch/_dynamo/__init__.py:140]: I'm saying that the docs here are confusing. You are indeed correct that my second question is what the code currently does. I did not understand that from the docs here.

File: torch/_dynamo/symbolic_convert.py [hunk 16]
```diff
@@ -471,6 +471,18 @@ def call_function(
             isinstance(x, VariableTracker)
             for x in itertools.chain(args, kwargs.values())
         )
+        inner_fn = None
+        if hasattr(fn, "value"):
+            inner_fn = fn.value
+        if hasattr(fn, "fn"):
+            inner_fn = fn.fn
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:478]: This is very suspicious. Why does VariableTracker not have a way to generally query attributes in this way...

File: torch/_dynamo/symbolic_convert.py [hunk 17]
```diff
@@ -1643,8 +1655,8 @@ def __init__(
         export,
         mutated_closure_cell_contents: Set[str],
     ):
-        super().__init__(
-            output=OutputGraph(f_globals, code_options, compiler_fn, self),
+        super(InstructionTranslator, self).__init__(
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:1658]: unintentional?

File: torch/fx/experimental/symbolic_shapes.py [hunk 18]
```diff
@@ -1031,29 +1041,33 @@ def create_unbacked_symint(self):
     # This is guaranteed to return a symbol or its negation is a sympy.Symbol,
     # but there may be a replacement that allows it to be immediately
     # simplified
-    def create_symbol(self, val: int, source: Source) -> "sympy.Expr":
+    def create_symbol(self, val: int, source: Source, dyn=False) -> "sympy.Expr":
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1044]: make this kwarg only?

File: torch/fx/experimental/symbolic_shapes.py [hunk 19]
```diff
@@ -1496,3 +1533,24 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
...
+    return not assume_static_by_default
+
+def _is_dim_dynamic(t, d):
+    return hasattr(t, "_dynamo_dynamic_indices") and d in t._dynamo_dynamic_indices
+
+def _is_int(expr):
+    try:
+        int(expr)
+    except RuntimeError:
+        return False
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1555]: This is terrible, can we do this test some other way. If expr is guaranteed to be sympy, you can just look at free symbols

File: torch/fx/experimental/symbolic_shapes.py [hunk 20]
```diff
@@ -1496,3 +1533,24 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
                 self.guards.append(
                     ShapeGuard(sympy.Eq(expr, concrete_val), stack))  # type: ignore[arg-type]
         return concrete_val
+
+def _should_allocate(user_marked_dynamic, assume_static_by_default):
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1537]: Make these bools kwarg only by default?

**Review by ezyang — APPROVED** [review]
please see comments

File: test/dynamo/test_misc.py (hunk 7 above)
Comment (voznesenskym) [reply test/dynamo/test_misc.py:4318]: True - Ostensibly you would mark dynamic ahead of time - we could also assert and manually request a .reset() - the way we do if a backend changes w/o a reset. Thoughts?

File: torch/_dynamo/__init__.py (hunk 11 above)
Comment (voznesenskym) [reply torch/_dynamo/__init__.py:140]: Let me try to rephrase this.

File: torch/_dynamo/symbolic_convert.py (hunk 16 above)
Comment (voznesenskym) [reply torch/_dynamo/symbolic_convert.py:478]: Legacy code, I could put up a refactor PR ahead of this to make `.value` the way.

File: torch/fx/experimental/symbolic_shapes.py (hunk 18 above)
Comment (voznesenskym) [reply torch/fx/experimental/symbolic_shapes.py:1044]: sure!

File: test/dynamo/test_misc.py (hunk 7 above)
Comment (ezyang) [reply test/dynamo/test_misc.py:4318]: There's no where to assert... except in the guard lol

File: torch/fx/experimental/symbolic_shapes.py (hunk 19 above)
Comment (voznesenskym) [reply torch/fx/experimental/symbolic_shapes.py:1555]: Not sure how to do this for when it is a SymInt...

File: test/dynamo/test_misc.py (hunk 7 above)
Comment (voznesenskym) [reply test/dynamo/test_misc.py:4318]: Let me hold off on this, and add a comment, I want to think about this more.

File: torch/fx/experimental/symbolic_shapes.py (hunk 19 above)
Comment (voznesenskym) [reply torch/fx/experimental/symbolic_shapes.py:1555]: Spoke offline ty for the pattern

**Comment by kit1980** [comment]
After this PR,  `test_autocast_sdpa_dynamic_shapes_static_default` started to fail with `RuntimeError: Cannot call sizes() on tensor with symbolic sizes/strides`: https://github.com/pytorch/pytorch/actions/runs/4206176846/jobs/7299657478

**Comment by kit1980** [comment]
> After this PR, `test_autocast_sdpa_dynamic_shapes_static_default` started to fail with `RuntimeError: Cannot call sizes() on tensor with symbolic sizes/strides`: https://github.com/pytorch/pytorch/actions/runs/4206176846/jobs/7299657478

Looks like the test was skipped on the PR because of some other issue that was later fixed.

**Comment by voznesenskym** [comment]
> test_autocast_sdpa_dynamic_shapes_static_default

Thanks, this is a coverage bug, we probably just need to exclude this test.

**Comment by voznesenskym** [comment]
> > After this PR, `test_autocast_sdpa_dynamic_shapes_static_default` started to fail with `RuntimeError: Cannot call sizes() on tensor with symbolic sizes/strides`: https://github.com/pytorch/pytorch/actions/runs/4206176846/jobs/7299657478
> 
> Looks like the test was skipped on the PR because of some other issue that was later fixed.

I wonder if for large PRs that change public API, we can forward fix?

**Comment by brendonydawson-oss** [comment]
Talk to me as I can help more than you can hack to make things work. Thanks for the hoe but we could do more
